library/core/src/main/java/com/google/android/exoplayer2/trackselection/TrackSelector.java

```diff
@@ -143,7 +143,7 @@ public abstract TrackSelectorResult selectTracks(
    * Calls {@link InvalidationListener#onTrackSelectionsInvalidated()} to invalidate all previously
    * generated track selections.
    */
-  protected final void invalidate() {
+  public final void invalidate() {
```

Comment:
Sorry about that.  If you want to deal with 'hot-plug' (which we do) you need to listen to the route changes and trigger trackselection when they occur.   There are no changes to the constraint based parameters ( `TrackSelectionParameters`) but the audio route will change will force a renderer/codec change.

The code we have in our application does this:

```java
    audioCapabilitiesReceiver = new AudioCapabilitiesReceiver(context, audioCapabilities -> {
      if (trackSelector != null) {
        trackSelector.invalidate();
      }
    });
    audioCapabilitiesReceiver.register();
```

We feel there are a few options here:

1. Open this method up and document why
1. Add a new parameter to `TrackSelectionParameters` that follows the API desired for when AudioDeviceInfo is available (for now, all it has to do is produce a change to the parameters to force an invalidate()
1. Incorporate this listener code down in `SimpleExoPlayer`, still this requires wider access to invalidate(), but with non documentation
1. Incorporate this listener code in the `trackselection` package, note this would pollute this package with retaining a reference to `Context`, which it currently wisely does not

We are ok with any of these choices, but wanted to allow the team to provide input before we added code to this pull request.  So, @ojw28 your thoughts?  You know `TrackSelection` the best.

Reply:
Triggering a track reselection is not enough. The `AudioCapabilities` of the `DefaultAudioSink` are currently immutable and need to be added a setter. That setter could send a `onCapabilityChange` notification to the renderer that could propagate a `onSupportedFormatChanged` or similar that would force a rerouting at the player level rather than been handled directly by the track selector.